From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SagarMakhar <sagarmakhar@gmail.com>
Date: Sun, 16 Jan 2022 08:25:59 +0000
Subject: [PATCH] BrightnessUtils: Conditionally use low gamma implementation
 for brightness curve

These devices report max brightness as 2047 or 4095

[fukiame: backport to Android 10]

Change-Id: I83b38a1334b8df156e4e53b06243c48d565be5d5
Co-authored-by: minaripenguin <minaripenguin@users.noreply.github.com>
Signed-off-by: Adithya R <gh0strider.2k18.reborn@gmail.com>
---
 .../android/settingslib/display/BrightnessUtils.java | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/packages/SettingsLib/src/com/android/settingslib/display/BrightnessUtils.java b/packages/SettingsLib/src/com/android/settingslib/display/BrightnessUtils.java
index 55723f9d..7bd3f8ee 100644
--- a/packages/SettingsLib/src/com/android/settingslib/display/BrightnessUtils.java
+++ b/packages/SettingsLib/src/com/android/settingslib/display/BrightnessUtils.java
@@ -16,11 +16,15 @@
 
 package com.android.settingslib.display;
 
+import android.os.SystemProperties;
 import android.util.MathUtils;
 
 public class BrightnessUtils {
 
-    public static final int GAMMA_SPACE_MAX = 1023;
+    private static final boolean sysUseLowGamma = SystemProperties.getBoolean(
+            "persist.sys.brightness.low.gamma", false);
+
+    public static final int GAMMA_SPACE_MAX = sysUseLowGamma ? 255 : 1023;
 
     // Hybrid Log Gamma constant values
     private static final float R = 0.5f;
@@ -51,6 +55,9 @@ public class BrightnessUtils {
      * @return The corresponding setting value.
      */
     public static final int convertGammaToLinear(int val, int min, int max) {
+        if (sysUseLowGamma) {
+            return MathUtils.constrain(val, min, max);
+        }
         final float normalizedVal = MathUtils.norm(0, GAMMA_SPACE_MAX, val);
         final float ret;
         if (normalizedVal <= R) {
@@ -87,6 +94,9 @@ public class BrightnessUtils {
      * @return The corresponding slider value
      */
     public static final int convertLinearToGamma(int val, int min, int max) {
+        if (sysUseLowGamma) {
+            return MathUtils.constrain(val, min, max);
+        }
         // For some reason, HLG normalizes to the range [0, 12] rather than [0, 1]
         final float normalizedVal = MathUtils.norm(min, max, val) * 12;
         final float ret;
